File: language/src/main/java/brj/nodes/ExecuteArrayNode.java
package brj.nodes;

import brj.BridjeLanguage;
import com.oracle.truffle.api.frame.VirtualFrame;
import com.oracle.truffle.api.nodes.ExecutableNode;
import com.oracle.truffle.api.nodes.ExplodeLoop;
import org.jetbrains.annotations.NotNull;

public class ExecuteArrayNode extends ExecutableNode {
    @Children
    private final ExecutableNode[] exprNodes;

    public ExecuteArrayNode(BridjeLanguage lang, ExecutableNode[] exprNodes) {
        super(lang);
        this.exprNodes = exprNodes;
    }

    @NotNull
    @Override
    @ExplodeLoop
    public Object[] execute(@NotNull VirtualFrame frame) {
        Object[] res = new Object[exprNodes.length];
        for (int i = 0; i < exprNodes.length; i++) {
            res[i] = exprNodes[i].execute(frame);
        }
        return res;
    }
}

File: language/src/main/kotlin/brj/nodes/ExecuteArrayNode.kt
package brj.nodes

import brj.BridjeNode
import com.oracle.truffle.api.frame.VirtualFrame
import com.oracle.truffle.api.nodes.ExplodeLoop
import com.oracle.truffle.api.source.SourceSection

class ExecuteArrayNode(
    @field:Children private val exprNodes: Array<BridjeNode>,
    loc: SourceSection? = null
) : BridjeNode(loc) {

    @ExplodeLoop
    override fun execute(frame: VirtualFrame): Array<Any> {
        val res = arrayOfNulls<Any>(exprNodes.size)
        for (i in exprNodes.indices) {
            res[i] = exprNodes[i].execute(frame)
        }
        @Suppress("UNCHECKED_CAST")
        return res as Array<Any>
    }
}
